### DISCUSSION TITLE
How to properly call "platform_hide" and "platform_show" on Linux?

### DISCUSSION THREAD

**Discussion by 4JX** [fltk-rs/fltk-rs#1183] [Q&A] [answered]
Unlike the Windows counterpart, the Linux logic calls `crate::app::flush();` when hiding and showing the window, which triggers the assert at:
https://github.com/fltk-rs/fltk-rs/blob/d59ab8063e0aa7d85c3bb8c969b9d7aaa3506943/fltk/src/app/visual.rs#L273-L276 
Meaning it panics when ran on a thread that is not the one the app was started in.

If I wanted to call the close and open functions from another thread, due to the requirements for [adding an entry](https://github.com/4JX/L5P-Keyboard-RGB/blob/ce7cb90a1637c38565997fa4943bb763dd288221/src/app.rs#L41) to a tray icon with the crate I'm using, how should I go about it? A channel with messages comes to mind, but would there be another option?

Also, some more platform specific behaviour happens at [this attempt](https://github.com/4JX/L5P-Keyboard-RGB/blob/e5c1bfa428d220eca611a163c0c842a89b71e1b7/src/gui/app.rs#L76-L78) to hide the window when the app is starting up, where running it on Windows hides it after a fast blink while Linux ignores it. So am I doing it incorrectly here?

Tested on `Arch Linux - Linux 5.17.1-arch1-1`

**Comment by MoAlyousef** [comment, marked as answer]
From my experimentation, platform_hide() doesn't work on wayland (or maybe particularly on kwin running wayland), so I wouldn't advise to use it generally. 

If you need the app to remain running even with the main window hidden, you can try invoking the event loop in a loop, using app::wait_for().
```rust
use fltk::{prelude::*, *};

fn main() {
    app::App::default().with_scheme(app::Scheme::Gtk);
    let mut win = window::Window::default().with_size(500, 300);
    let mut btn = button::Button::new(160, 200, 80, 30, "Hide");
    win.show();
    btn.set_callback(move |_| win.hide());
    loop {
        app::wait_for(1.0).unwrap();
    }
}
```
Window::hide() will also run the assert that it's called from the ui thread.  The only safe way around it is via using channels:
```rust
use fltk::{prelude::*, *};

fn main() {
    let (s, r) = app::channel();
    app::App::default().with_scheme(app::Scheme::Gtk);
    let mut win = window::Window::default().with_size(500, 300);
    let mut btn = button::Button::new(160, 200, 80, 30, "Hide");
    win.show();
    btn.set_callback(move |_| {
        let s = s.clone();
        std::thread::spawn(move || {
            std::thread::sleep(std::time::Duration::from_secs(2));
            s.send(1);
        });
    });
    loop {
        app::wait_for(1.0).unwrap();
        if let Some(msg) = r.recv() {
            if msg == 1 {
                win.hide();
            }
        }
    }
}
```

You can move the message handler into an add_idle method callback:
```rust
    app::add_idle3(move |_| {
        if let Some(msg) = r.recv() {
            if msg == 1 {
                win.hide();
            }
        }
    });
```

An unsafe way of doing things is to do what plarform_show and platform_hide do things, via calling the respective XMapWindow and XUnmapWindow, you would need to also call XFlush after either.

Regarding:
```rust
if !show_window {
	win.platform_hide()
};
```
You can use the app::wait_for() inside a loop, and this will run the event loop even without a window, so the logic for showing/hiding the window can be moved to inside the condition:
```rust
if show_window {
        let mut win = Self::create_window(manager);
	win.set_callback(|win| win.platform_hide());
        app::add_idle3(move |_| {
            if let Some(msg) = r.recv() {
                if msg == 1 {
                    win.hide();
                }
            }
        });
};
// code
tray.add_menu_item("Show", move || s.send(1)).unwrap();
```
